tests/Unit/EmogrifierTest.php

```diff
@@ -690,6 +690,143 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+        // like 'color: red;' or 'color: blue;'.
+        return [$selector . ' { %1$s }', $matchedTagPart . ' style="%1$s">'];
+    }
+
+    /**
+     * @param string $firstSelector A selector expression to be used in an earlier CSS declaration
+     * @param string $secondSelector Some other selector expression to be used in a later CSS declaration
+     * @param bool $firstSelectorIsMoreSpecific Whether the first selector's rules are expected to prevail
+     * @param string $matchedTagPart Tag (in {@see EmogrifierTest::COMMON_TEST_HTML}) expected to be matched by both
+     *                               selectors, without the closing '>', e.g. '<p class="p-1"'
+     *
+     * @return string[][]
+     */
+    private function getSelectorSpecificityCssDataSet(
+        $firstSelector,
+        $secondSelector,
+        $firstSelectorIsMoreSpecific,
+        $matchedTagPart
+    ) {
+        // The sprintf placeholders %1$s and %2$s will automatically be replaced with CSS declarations
+        // like 'color: red;' or 'color: blue;'.
+        return [
+            $firstSelector . ' { %1$s } ' . $secondSelector . ' { %2$s }',
+            $matchedTagPart . ($firstSelectorIsMoreSpecific ? ' style="%1$s">' : ' style="%2$s">')
+        ];
+    }
+
+    /**
+     * @param string $name Base name for identifying datasets in test results
+     * @param string $firstSelector A selector expression
+     * @param string $secondSelector Some other selector expression
+     * @param string $operator '<', '==' or '>', defining whether the first selector is expected to have, respectively,
+     *                         lower, equal or higher specificity than the second
+     * @param string $matchedTagPart Tag (in {@see EmogrifierTest::COMMON_TEST_HTML}) expected to be matched by both
+     *                               selectors, without the closing '>', e.g. '<p class="p-1"'
+     *
+     * @return string[][]
+     */
+    private function getSelectorSpecificityCssDataSets(
+        $name,
+        $firstSelector,
+        $secondSelector,
+        $operator,
+        $matchedTagPart
+    ) {
+        return [
+            $name . ' (1st selector valid for test)' => $this->getMatchedCssDataSet($firstSelector, $matchedTagPart),
+            $name . ' (2nd selector valid for test)' => $this->getMatchedCssDataSet($secondSelector, $matchedTagPart),
+            // Check first selector matches when its rule is last unless the second has higher specificity
+            $name . ' (1st selector in last rule)' => $this->getSelectorSpecificityCssDataSet(
+                $secondSelector,
+                $firstSelector,
+                $operator === '<',
+                $matchedTagPart
+            ),
+            $name . ' (2nd selector in last rule)' => $this->getSelectorSpecificityCssDataSet(
+                $firstSelector,
+                $secondSelector,
+                $operator === '>',
+                $matchedTagPart
```

Comment:
We can simplify this by having 2 or 3 separate tests (with shared or separate data providers): one for selectors that have the same specificity (what a word!), and one or two for different specificities (depending on whether you think that we order matters).

What do you think?

Reply:
Have refactored along those lines.